zellij-utils/src/windows_utils/named_pipe.rs

```diff
@@ -0,0 +1,326 @@
+        let value = $call;
+        if value != INVALID_HANDLE_VALUE {
+            Ok(value)
+        } else {
+            Err(std::io::Error::last_os_error())
+        }
+    }};
+}
+
+#[repr(transparent)]
+struct EventedOverlapped(OVERLAPPED);
+
+impl Drop for EventedOverlapped {
+    fn drop(&mut self) {
+        if (!self.0.hEvent.is_null()) {
+            unsafe { CloseHandle(self.0.hEvent); }
+        }
+    }
+}
+
+/// Helper function to create an instance of [OVERLAPPED] with a new unique event
+fn create_overlapped_with_new_event() -> io::Result<EventedOverlapped> {
+    let mut overlapped = create_zeroed_overlapped();
+    overlapped.hEvent = {
+        let value = unsafe { CreateEventW(ptr::null_mut(), TRUE, TRUE, ptr::null_mut()) };
+        if !value.is_null() {
+            Ok(value)
+        } else {
+            Err(std::io::Error::last_os_error())
+        }
+    }?;
+
+    Ok(EventedOverlapped(overlapped))
+}
+
+/// Helper function to create an zeroed instance of [OVERLAPPED]
+fn create_zeroed_overlapped() -> OVERLAPPED {
+    // SAFETY: Docs state to use an OVERLAPPED-Struct with all Members zeroed
+    unsafe { MaybeUninit::zeroed().assume_init() }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Pipe {
+    pipe_name: Arc<[u16]>,
+}
+
+struct PipeAcceptIterator {
+    pipe: Pipe,
+}
+
+impl Pipe {
+    pub fn new(name: impl AsRef<OsStr>) -> Self {
+        let pipe_name = Pipe::convert_pipe_name(name.as_ref());
+        Self {
+            pipe_name: Arc::from(pipe_name),
+        }
+    }
+
+    pub fn incoming(&self) -> impl Iterator<Item = io::Result<PipeStream>> {
+        PipeAcceptIterator { pipe: self.clone() }
+    }
+
+    pub fn connect(&self) -> io::Result<PipeStream> {
+        loop {
+            let client: Result<HANDLE, io::Error> = call_with_last_error!(unsafe {
+                CreateFileW(
+                    self.pipe_name.as_ptr(),
+                    GENERIC_READ | GENERIC_WRITE,
+                    FILE_SHARE_READ | FILE_SHARE_WRITE,
+                    ptr::null_mut(),
+                    OPEN_EXISTING,
+                    FILE_FLAG_OVERLAPPED,
+                    ptr::null_mut(),
+                )
+            });
+
+            match client {
+                Ok(handle) => break Ok(handle.into()),
+                Err(err) if err.raw_os_error() == Some(ERROR_PIPE_BUSY as i32) => {
+                    continue;
+                },
+                // TODO check when a call to WaitNamedPipe is usefull
+                Err(err) => return Err(err),
+            };
+        }
+    }
+
+    pub fn accept(&self) -> io::Result<PipeStream> {
+        let server_listener_pipe_handle = unsafe {
+            CreateNamedPipeW(
+                self.pipe_name.as_ptr(),                   // pipe name
+                PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED, // read/write access
+                PIPE_TYPE_BYTE |       // message type pipe
+                    PIPE_WAIT, // blocking mode
```

Comment:
How does `PIPE_WAIT` and `FILE_FLAG_OVERLAPPED` interact in this context.

Reply:
The create named pipe docs also say `PIPE_NOWAIT` is for legacy compatibility.